python/ray/data/grouped_data.py

```diff
@@ -297,19 +297,44 @@ def map_groups(
         else:
             sorted_ds = self._dataset.repartition(1)
 
-        # Returns the group boundaries.
-        def get_key_boundaries(block_accessor: BlockAccessor):
+        def get_key_boundaries(block_accessor: BlockAccessor) -> List[int]:
+            """Compute block boundaries based on the key(s)"""
+
             import numpy as np
 
-            boundaries = []
             # Get the keys of the batch in numpy array format
             keys = block_accessor.to_numpy(self._key)
-            start = 0
-            while start < keys.size:
-                end = start + np.searchsorted(keys[start:], keys[start], side="right")
-                boundaries.append(end)
-                start = end
-            return boundaries
+
+            has_single_key = isinstance(keys, np.ndarray)
+
+            if has_single_key:
+                boundaries = []
+                start = 0
+                while start < len(keys):
+                    end = start + np.searchsorted(
+                        keys[start:], keys[start], side="right"
+                    )
+                    boundaries.append(end)
+                    start = end
+                return boundaries
+
+            else:
+                # For multiple keys, we generate a separate tuple column
+                # Due to how np functions treats tuples, we need to cast
+                # it into an array of tuple
+                first_key = next(iter(keys))
+                arr = np.empty(len(keys[first_key]), dtype=object)
+                arr[:] = list(zip(*keys.values()))
+
+                boundaries = []
+                start = 0
+                while start < len(arr):
+                    x = np.empty(1, dtype=object)
+                    x[0] = arr[start]
```

Comment:
To avoid duplication and minimize complexity, is there any way to write this so that also works for the single key case?

Reply:
Yes, using a custom class instead of `tuple` will simplify a bit. numpy treats them as `object` anyway